opm/autodiff/BlackoilModel_impl.hpp

```diff
@@ -0,0 +1,2247 @@
+  Copyright 2015 IRIS AS
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef OPM_BLACKOILMODEL_IMPL_HEADER_INCLUDED
+#define OPM_BLACKOILMODEL_IMPL_HEADER_INCLUDED
+
+#include <opm/autodiff/BlackoilModel.hpp>
+
+#include <opm/autodiff/AutoDiffBlock.hpp>
+#include <opm/autodiff/AutoDiffHelpers.hpp>
+#include <opm/autodiff/GridHelpers.hpp>
+#include <opm/autodiff/BlackoilPropsAdInterface.hpp>
+#include <opm/autodiff/GeoProps.hpp>
+#include <opm/autodiff/WellDensitySegmented.hpp>
+#include <opm/autodiff/WellStateFullyImplicitBlackoil.hpp>
+
+#include <opm/core/grid.h>
+#include <opm/core/linalg/LinearSolverInterface.hpp>
+#include <opm/core/linalg/ParallelIstlInformation.hpp>
+#include <opm/core/props/rock/RockCompressibility.hpp>
+#include <opm/core/simulator/BlackoilState.hpp>
+#include <opm/core/utility/ErrorMacros.hpp>
+#include <opm/core/utility/Exceptions.hpp>
+#include <opm/core/utility/Units.hpp>
+#include <opm/core/well_controls.h>
+#include <opm/core/utility/parameters/ParameterGroup.hpp>
+
+#include <cassert>
+#include <cmath>
+#include <iostream>
+#include <iomanip>
+#include <limits>
+//#include <fstream>
+
+// A debugging utility.
+#define DUMP(foo)                                                       \
+    do {                                                                \
+        std::cout << "==========================================\n"     \
+                  << #foo ":\n"                                         \
+                  << collapseJacs(foo) << std::endl;                    \
+    } while (0)
+
+#define DUMPVAL(foo)                                                    \
+    do {                                                                \
+        std::cout << "==========================================\n"     \
+                  << #foo ":\n"                                         \
+                  << foo.value() << std::endl;                          \
+    } while (0)
+
+#define DISKVAL(foo)                                                    \
+    do {                                                                \
+        std::ofstream os(#foo);                                         \
+        os.precision(16);                                               \
+        os << foo.value() << std::endl;                                 \
+    } while (0)
```

Comment:
Since some people are using them I will instead rename them by adding a OPM_AD_ prefix in front of each.